pkg/vcs: export GitWrapper

a-nogikh opened the pull request:
One more part of #5620 (see the discussion there as well).
***
It will let other parts of the code to use only the git-specific functionality.

tarasmadan commented:
Commit comment doesn't explain what's going on.
Please make it valuable for the future readers.

tarasmadan commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,144 @@ func ParseGitDiff(patch []byte) []string {
 	return files
 }
+
+type Git struct {
```
Why public? Because following PRs need it?

tarasmadan replied:
Got it now. We have a git command line tool wrapper and the git repo abstraction now.

tarasmadan commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,144 @@ func ParseGitDiff(patch []byte) []string {
+}
+
+// Reset resets the git repo to a known clean state.
+func (git Git) Reset() error {
```
Some of the functions receive pointers. Some values.

tarasmadan commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,144 @@ func ParseGitDiff(patch []byte) []string {
+	return osutil.Run(3*time.Hour, cmd)
+}
+
+func (git Git) gitCommand(args ...string) (*exec.Cmd, error) {
```
s/gitCommand/command/?

a-nogikh replied:
Renamed

tarasmadan commented:
Please start from the explanation why this git wrapper is needed.

a-nogikh commented:
Updated the commit description to

```
    pkg/vcs: export GitWrapper
    
    The current common vcs interface and its implementations are designed to
    support a very big number of situations (git modules, cleaning up
    complex state, etc), which is too heavy and restrictive when we just
    need a thin wrapper over the git command that supports a few basic
    operations.
    
    Refactor pkg/vcs to split out the wrapper code from the big git
    implementation of the vcs.Repo interface.
    
    Export the wrapper to enable reuse in other parts of the system.

```

Re. the usages -- see `syz-cluster/pkg/triage/git.go` in #5620.

Specifically, I could not use `pkg/vcs`'s generic `Patch` because it would not support raw emails (while `git apply` does), so I needed a wrapper over the git command. The wrapper code was already incorporated into `pkg/vcs`'s `git`, so instead of duplicating the code I split it out from the `git` implementation.

a-nogikh commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,144 @@ func ParseGitDiff(patch []byte) []string {
+}
+
+// Reset resets the git repo to a known clean state.
+func (git Git) Reset() error {
```
I used it rather to signal the intent that the methods are not modifying the state of the `Git` object. I hope that the Go compiler may optimize it and not do copying in this case or, even if it still does copy, it costs almost nothing since the object is very small.
